### PR TITLE
Feature/themes

### PR DESCRIPTION
**Pull request by TadeasKriz**

### CODE REVIEW

File: Example/Source/Component/Playground/ExampleView2.ui.xml [hunk 1]
```diff
@@ -14,20 +14,18 @@
...
             numberOfLines="0"
             textAlignment="center"
-            layout:top="[portrait] super inset(10);
-                        [landscape] super inset(0)"
+            layout:top="super inset(10)"
             layout:centerX="super"
             layout:leading=":gte super inset(10)"
             layout:trailing=":lte super inset(10)">
-            <attributedText>
+            <attributedText foregroundColor="theme.textOnOverlay">
```
Comment (MatyasKriz) [inline Example/Source/Component/Playground/ExampleView2.ui.xml:28]: It's awesome that we can even do this.

File: Example/Source/Component/Playground/ExampleView2.ui.xml [hunk 2]
```diff
@@ -14,20 +14,18 @@
...
+        layout:edges="super">
 
         <Label
             layout:id="label"
             font=":regular@20"
             numberOfLines="0"
             textAlignment="center"
-            layout:top="[portrait] super inset(10);
-                        [landscape] super inset(0)"
+            layout:top="super inset(10)"
```
Comment (MatyasKriz) [inline Example/Source/Component/Playground/ExampleView2.ui.xml:24]: Why did you remove the conditions? I reckon that Playground should test all the micro-functionality we offer.

File: Sources/Tokenizer/Contexts/SupportedPropertyTypeContext.swift [hunk 3]
```diff
@@ -8,29 +8,29 @@
...
+        self.parentContext = parentContext
         self.value = value
     }
 }
 
 extension SupportedPropertyTypeContext: HasParentContext {
-    public var parentContext: PropertyContext {
-        return propertyContext
-    }
+//    public var parentContext: PropertyContext {
```
Comment (MatyasKriz) [inline Sources/Tokenizer/Contexts/SupportedPropertyTypeContext.swift:21]: x_x

File: Sources/Tokenizer/Elements/Implementation/AttributedText.swift [hunk 4]
```diff
@@ -39,9 +39,24 @@ public struct AttributedText: ElementSupportedPropertyType {
     public let localProperties: [Property]
     public let parts: [AttributedText.Part]
 
+    public var requiresTheme: Bool {
+        return localProperties.contains(where: { $0.anyValue.requiresTheme }) ||
```
Comment (MatyasKriz) [inline Sources/Tokenizer/Elements/Implementation/AttributedText.swift:43]: Is there a reason not to use trailing closures instead?

File: Sources/Tokenizer/Themes/ApplicationDescription.swift [hunk 5]
```diff
@@ -0,0 +1,120 @@
+//
+//  HyperdriveApp.swift
+//  Example
```
Comment (MatyasKriz) [inline Sources/Tokenizer/Themes/ApplicationDescription.swift:3]: `ReactantUI` perhaps?

File: Sources/reactant-ui/GenerateCommand.swift [hunk 6]
```diff
@@ -73,6 +80,19 @@ class GenerateCommand: Command {
...
         }
 
+        // ApplicationDescription is not required. We can work with default values and it makes it backward compatible.
+        let applicationDescription: ApplicationDescription
+        let applicationDescriptionPath = applicationDescriptionFile.value
+        if let applicationDescriptionPath = applicationDescriptionPath {
+            let applicationDescriptionData = try Data(contentsOf: URL(fileURLWithPath: applicationDescriptionPath))
+            let xml = SWXMLHash.parse(applicationDescriptionData)
+            // FIXME Ugly force unwrapping
+            let node = xml["Application"].element!
```
Comment (MatyasKriz) [inline Sources/reactant-ui/GenerateCommand.swift:90]: `singleOrNoElement` instead?

File: Sources/reactant-ui/GenerateCommand.swift [hunk 7]
```diff
@@ -192,13 +218,119 @@ class GenerateCommand: Command {
...
+        } else {
+            rxSwiftShim = ""
+        }
+
+        return """
+
+        public enum ApplicationTheme: String, ReactantThemeDefinition {
+            public static let selector = ReactantThemeSelector<ApplicationTheme>(defaultTheme: .\(description.defaultTheme))
+
+        \(cases)
```
Comment (MatyasKriz) [inline Sources/reactant-ui/GenerateCommand.swift]: I assume the texts themselves are indented instead. Is that wise?

**Review by MatyasKriz — APPROVED** [review]
Great addition to ReactantUI. Good job.

File: Example/Source/Component/Playground/ExampleView2.ui.xml (hunk 2 above)
Comment (TadeasKriz) [reply Example/Source/Component/Playground/ExampleView2.ui.xml:24]: Thanks for noticing, totally forgot about it. There's probably something wrong when running the conditions on a device, because it wasn't displaying properly before removing the conditions.

File: Sources/Tokenizer/Elements/Implementation/AttributedText.swift (hunk 4 above)
Comment (TadeasKriz) [reply Sources/Tokenizer/Elements/Implementation/AttributedText.swift:43]: In my mind it's consistency with the usage of the same method in conditions. There you have to use `(where: { .. })` and I feel that this makes it consistent.

File: Sources/reactant-ui/GenerateCommand.swift (hunk 6 above)
Comment (TadeasKriz) [reply Sources/reactant-ui/GenerateCommand.swift:90]: There's no such thing on `XMLIndexer`

File: Sources/reactant-ui/GenerateCommand.swift (hunk 7 above)
Comment (TadeasKriz) [reply Sources/reactant-ui/GenerateCommand.swift]: Well. Let's not look at the generated swift code. This is something we should solve as a separate PR.

Comment (MatyasKriz) [reply Sources/reactant-ui/GenerateCommand.swift]: Why is that?

Comment (TadeasKriz) [reply Sources/reactant-ui/GenerateCommand.swift]: This is not inside the `UIGenerator`, but in `GenerateCommand` where it uses simple `var output: [String]` and `output.append(...)`. Because of that, there's not even basic auto-indentation. The indentation in `UIGenerator` is done by the `l` function.

File: Example/Source/Component/Playground/ExampleView2.ui.xml (hunk 1 above)
Comment (TadeasKriz) [reply Example/Source/Component/Playground/ExampleView2.ui.xml:28]: That's thanks to the `Property` and `Context` APIs, so awesome.
